hydra-node/test/Hydra/JSONSchema.hs

```diff
@@ -0,0 +1,116 @@
+      -- the specification. Note this, because this uses an external tool each
+      -- property iteration is pretty slow. So instead, we run the property only
+      -- once, but on a list of 100 elements all arbitrarily generated.
+      [ forAllShrink (vectorOf 100 arbitrary) shrink $ \(a :: [a]) ->
+          monadicIO $
+            do
+              run ensureSystemRequirements
+              let obj = Aeson.encode a
+              (exitCode, _out, err) <- run $ do
+                writeFileLBS inputFile obj
+                readProcessWithExitCode "jsonschema" ["-i", inputFile, specFile] mempty
+              monitor $ counterexample err
+              monitor $ counterexample (decodeUtf8 obj)
+              assert (exitCode == ExitSuccess)
+      , -- This second sub-property ensures that any key found in the
+        -- specification corresponds to a constructor in the corresponding
+        -- data-type. This in order the document in sync and make sure we don't
+        -- left behind constructors which no longer exists.
+        forAllBlind (vectorOf 1000 arbitrary) $
+          \(a :: [a]) -> monadicIO $ do
+            specs <- run $ Aeson.decodeFileStrict specFile
+            let unknownConstructors = Map.keys $ Map.filter (== 0) $ classify specs a
+            unless (null unknownConstructors) $ do
+              let commaSeparated = intercalate ", " (toString <$> unknownConstructors)
+              monitor $ counterexample $ "Unimplemented constructors present in specification: " <> commaSeparated
+              assert False
+      ]
+ where
+  -- Like Generics, if you squint hard-enough.
+  strawmanGetConstr :: a -> Text
+  strawmanGetConstr = toText . Prelude.head . words . show
+
+  classify :: Maybe Aeson.Value -> [a] -> Map Text Integer
+  classify (Just specs) =
+    let knownKeys =
+          case specs ^? key "properties" . key namespace . key "items" . key "oneOf" . _Array of
+            Just (toList -> es) ->
+              let ks = mapMaybe (\(e :: Aeson.Value) -> e ^? key "title" . _String) es
+               in Map.fromList $ zip ks (repeat @Integer 0)
+            _ ->
+              mempty
+
+        countMatch (strawmanGetConstr -> tag) =
+          Map.alter (Just . maybe 1 (+ 1)) tag
+     in foldr countMatch knownKeys
+  classify _ =
+    error $ "Invalid specification file. Does not decode to an object: " <> show specFile
+
+-- | Prepare the environment (temp directory) with the JSON specification. We
+-- maintain a YAML version of a JSON-schema, for it is more convenient to write.
+-- But tools (and in particular jsonschema) only works from JSON, so this
+-- function makes sure to also convert our local yaml into JSON.
+withJsonSpecifications ::
+  FilePath ->
+  ((FilePath, FilePath) -> IO ()) ->
+  IO ()
+withJsonSpecifications specFile action = do
+  specs <- Yaml.decodeFileThrow @_ @Aeson.Value =<< Pkg.getDataFileName specFile
```

Comment:
I don't know why this works for `api-log.yaml` since it's not specified in the `data-files` in `.cabal` :shrug:

Reply:
Cabal rox